collector/infiniband.go

```diff
@@ -0,0 +1,166 @@
+// Copyright 2017 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build linux
+// +build !noinfiniband
+
+package collector
+
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/log"
+)
+
+type infinibandCollector struct {
+	metricDescs              map[string]*prometheus.Desc
+	infinibandCounters       map[string]string
+	infinibandPath           string
+	infinibandNoDevicesFound error
+	infinibandNoPortsFound   error
+}
+
+func init() {
+	Factories["infiniband"] = NewInfiniBandCollector
+}
+
+// Takes a prometheus registry and returns a new Collector expossing InfiniBand statistics.
+func NewInfiniBandCollector() (Collector, error) {
+	var i infinibandCollector
+
+	// Filenames of all InfiniBand counter metrics including a detailed description.
+	i.infinibandCounters = map[string]string{
+		"link_downed":            "Number of times the link failed to recover from an error state and went down",
+		"link_error_recovery":    "Number of times the link successfully recovered from an error state",
+		"multicast_rcv_packets":  "Number of multicast packets received (including errors)",
```

Comment:
I definitely agree with you that spelling out "receive" would be beneficial, but `multicast_rcv_packets` is the name of the file I read the metric from, hence the use here. I extract every metric's filename from the key in this map.